### PR TITLE
PLUGIN-78 - Zendesk Batch Source Multi Object plugin

### PR DESCRIPTION
**Pull request by rostyslavmorozov**
WIKI: https://wiki.cask.co/display/CE/Zendesk+plugins
JIRA: https://issues.cask.co/browse/PLUGIN-78

Included:
- Multi Object Batch Source

### CODE REVIEW

File: docs/ZendeskBatchMultiSource-batchsource.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+**Reference Name:** Name used to uniquely identify this source for lineage, annotating metadata, etc.
+
+**Admin Email:** Zendesk admin email.
+
+**API Token:** Zendesk API token. Can be obtained from the Zendesk Support Admin interface.
+Check out [Zendesk documentation](https://support.zendesk.com/hc/en-us/articles/226022787-Generating-a-new-API-token-)
+for API Token generation.
+
+**Subdomains:** List of Zendesk Subdomains to read object from.
```
Comment (aonischuk) [inline docs/ZendeskBatchMultiSource-batchsource.md]: "to read **the** object**s** from"

File: docs/ZendeskBatchMultiSource-batchsource.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+**Objects to Pull:** Objects to pull from Zendesk API. Supports multi-select.
+
+**Objects to Skip:** Objects to skip from Zendesk API. Supports multi-select.
+
+**Start Date:** Filter data to only include records where Zendesk modified date is greater than
```
Comment (aonischuk) [inline docs/ZendeskBatchMultiSource-batchsource.md]: Filter data to include only records which have Zendesk modified date greater ...

File: docs/ZendeskBatchMultiSource-batchsource.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+|              Format              |       Format Syntax       |          Example          |
+| -------------------------------- | ------------------------- | ------------------------- |
+| Date, time, and time zone offset | YYYY-MM-DDThh:mm:ss+hh:mm | 1999-01-01T23:01:01+01:00 |
+|                                  | YYYY-MM-DDThh:mm:ss-hh:mm | 1999-01-01T23:01:01-08:00 |
+|                                  | YYYY-MM-DDThh:mm:ssZ      | 1999-01-01T23:01:01Z      |
+
+Start Date is required for batch objects like: Ticket Comments, Organizations, Ticket Metric Events, Tickets, Users.
+
+**End Date:** Filter data to only include records where Zendesk modified date is less than
```
Comment (aonischuk) [inline docs/ZendeskBatchMultiSource-batchsource.md]: same

File: docs/ZendeskBatchMultiSource-batchsource.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+|              Format              |       Format Syntax       |          Example          |
+| -------------------------------- | ------------------------- | ------------------------- |
+| Date, time, and time zone offset | YYYY-MM-DDThh:mm:ss+hh:mm | 1999-01-01T23:01:01+01:00 |
+|                                  | YYYY-MM-DDThh:mm:ss-hh:mm | 1999-01-01T23:01:01-08:00 |
+|                                  | YYYY-MM-DDThh:mm:ssZ      | 1999-01-01T23:01:01Z      |
+
+Specifying this along with `Start Date` allows reading data modified within a specific time window. 
+If no value is provided, no upper bound is applied.
+
+**Satisfaction Ratings Score:** Score filter for Satisfaction Ratings object.
```
Comment (aonischuk) [inline docs/ZendeskBatchMultiSource-batchsource.md]: It's hard to understand what this is from this description.

File: src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskBatchMultiSource.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  public ZendeskBatchMultiSource(ZendeskBatchSourceConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    FailureCollector failureCollector = pipelineConfigurer.getStageConfigurer().getFailureCollector();
+    validateConfiguration(failureCollector);
+
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(null);
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskBatchMultiSource.java]: why do we clear schema during validation. Is this something intentional?

File: src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskBatchMultiSource.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+      config, config.getObjects(), schemasStrings)));
+  }
+
+  @Override
+  public void transform(KeyValue<NullWritable, StructuredRecord> input,
+                        Emitter<StructuredRecord> emitter) throws Exception {
+    emitter.emit(input.getValue());
+  }
+
+  private void validateConfiguration(FailureCollector failureCollector) {
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskBatchMultiSource.java]: Do we need a separate method for only two lines of code?

File: src/main/java/io/cdap/plugin/zendesk/source/common/config/BaseZendeskSourceConfig.java [hunk 7]
```diff
@@ -87,12 +118,33 @@ public String getApiToken() {
...
+      .collect(Collectors.toList());
+  }
+
   public void validate(FailureCollector collector) {
     IdUtils.validateReferenceName(referenceName, collector);
-    if (!EmailValidator.getInstance().isValid(adminEmail)) {
-      collector.addFailure(String.format("Email '%s' is invalid.", adminEmail), null)
+    if (!containsMacro(PROPERTY_ADMIN_EMAIL)
+      && !EmailValidator.getInstance().isValid(adminEmail)) {
+      collector.addFailure(String.format("Provided email is invalid: %s", adminEmail), null)
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/zendesk/source/common/config/BaseZendeskSourceConfig.java]: '%s' is not a valid email.

File: widgets/ZendeskBatchMultiSource-batchsource.json [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+          "label": "Reference Name",
+          "name": "referenceName"
+        },
+        {
+          "widget-type": "textbox",
+          "label": "Admin Email",
+          "name": "adminEmail"
+        },
+        {
+          "widget-type": "textbox",
```
Comment (aonischuk) [inline widgets/ZendeskBatchMultiSource-batchsource.json]: is this secret info or not? If it is we should use "password" widget-type.

**Review by aonischuk — CHANGES_REQUESTED** [review]

File: docs/ZendeskBatchMultiSource-batchsource.md (hunk 1 above)
Comment (rostyslavmorozov) [reply docs/ZendeskBatchMultiSource-batchsource.md]: Done

File: docs/ZendeskBatchMultiSource-batchsource.md (hunk 2 above)
Comment (rostyslavmorozov) [reply docs/ZendeskBatchMultiSource-batchsource.md]: Done

File: docs/ZendeskBatchMultiSource-batchsource.md (hunk 3 above)
Comment (rostyslavmorozov) [reply docs/ZendeskBatchMultiSource-batchsource.md]: Done

File: docs/ZendeskBatchMultiSource-batchsource.md (hunk 4 above)
Comment (rostyslavmorozov) [reply docs/ZendeskBatchMultiSource-batchsource.md]: Done

File: src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskBatchMultiSource.java (hunk 5 above)
Comment (rostyslavmorozov) [reply src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskBatchMultiSource.java]: Done

File: src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskBatchMultiSource.java (hunk 6 above)
Comment (rostyslavmorozov) [reply src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskBatchMultiSource.java]: Done

File: src/main/java/io/cdap/plugin/zendesk/source/common/config/BaseZendeskSourceConfig.java (hunk 7 above)
Comment (rostyslavmorozov) [reply src/main/java/io/cdap/plugin/zendesk/source/common/config/BaseZendeskSourceConfig.java]: Done

File: widgets/ZendeskBatchMultiSource-batchsource.json (hunk 8 above)
Comment (rostyslavmorozov) [reply widgets/ZendeskBatchMultiSource-batchsource.json]: Done

**Review by aonischuk — APPROVED** [review]

**Pull request merged**